File: server/hotrod/src/test/scala/org/infinispan/server/hotrod/test/HotRodMagicKeyGenerator.scala
package org.infinispan.server.hotrod.test

import org.infinispan.Cache
import java.util.Random
import org.infinispan.marshall.core.JBossMarshaller

/**
 * {@link org.infinispan.distribution.MagicKey} equivalent for HotRod
 *
 * @author Galder Zamarreño
 * @since 5.2
 */
object HotRodMagicKeyGenerator {

   def newKey(cache: Cache[AnyRef, AnyRef]): Array[Byte] = {
      val ch = cache.getAdvancedCache.getDistributionManager.getReadConsistentHash
      val nodeAddress = cache.getAdvancedCache.getRpcManager.getAddress
      val r = new Random()

      val sm = new JBossMarshaller()
      for (i <- 0 to 1000) {
         val candidate = String.valueOf(r.nextLong())
         val candidateBytes = sm.objectToByteBuffer(candidate, 64)
         if (ch.isKeyLocalToNode(nodeAddress, candidateBytes)) {
            return candidateBytes
         }
      }

      throw new RuntimeException("Unable to find a key local to node " + cache)
   }

   def getStringObject(bytes: Array[Byte]): String = {
      val sm = new JBossMarshaller()
      sm.objectFromByteBuffer(bytes).asInstanceOf[String]
   }

}

File: server/hotrod/src/test/java/org/infinispan/server/hotrod/test/HotRodMagicKeyGenerator.java
package org.infinispan.server.hotrod.test;

import java.util.Random;

import org.infinispan.Cache;
import org.infinispan.jboss.marshalling.commons.GenericJBossMarshaller;
import org.infinispan.distribution.LocalizedCacheTopology;

/**
 * {@link org.infinispan.distribution.MagicKey} equivalent for HotRod
 *
 * @author Galder Zamarreño
 * @since 5.2
 */
public class HotRodMagicKeyGenerator {

   public static byte[] newKey(Cache<?, ?> cache) throws Exception {
      LocalizedCacheTopology cacheTopology = cache.getAdvancedCache().getDistributionManager().getCacheTopology();
      Random r = new Random();

      GenericJBossMarshaller sm = new GenericJBossMarshaller();
      for (int i = 0; i < 1000; i++) {
         String candidate = String.valueOf(r.nextLong());
         byte[] candidateBytes = sm.objectToByteBuffer(candidate, 64);
         if (cacheTopology.isReadOwner(candidateBytes)) {
            return candidateBytes;
         }
      }

      throw new RuntimeException("Unable to find a key local to node " + cache);
   }

   public static String getStringObject(byte[] bytes) throws Exception {
      GenericJBossMarshaller sm = new GenericJBossMarshaller();
      return ((String) sm.objectFromByteBuffer(bytes));
   }
}
